cpp/daal/src/algorithms/dtrees/forest/oneapi/cl_kernels/df_tree_level_build_helper_kernels.cl

```diff
@@ -41,33 +41,83 @@ DECLARE_SOURCE(
         treeOrder[offset + id] = treeOrderBuf[offset + id];
     }
 
-    __kernel void doLevelPartition(const __global int * data, const __global int * nodeList, const __global int * treeOrder,
-                                   __global int * treeOrderBuf, int nFeatures) {
-        const int nNodeProp  = NODE_PROPS; // num of split attributes for node
-        const int leafMark   = -1;
-        const int local_id   = get_sub_group_local_id();
-        const int nodeId     = get_global_id(1);
-        const int local_size = get_sub_group_size();
+    __kernel void doLevelPartitionByGroups(const __global int * data, const __global int * nodeList, __global int * nodeAuxList,
+                                           const __global int * treeOrder, __global int * treeOrderBuf, int nNodes, int nFeatures) {
+        const int nNodeProp    = NODE_PROPS;     // num of split attributes for node
+        const int nAuxNodeProp = AUX_NODE_PROPS; // num of auxilliary attributes for node
+        const int maxBlocksNum = PARTITION_MAX_BLOCKS_NUM;
+        const int minBlockSize = PARTITION_MIN_BLOCK_SIZE;
+        const int leafMark     = -1;
+
+        const int sub_group_size               = get_sub_group_size();
+        const int work_group_size              = get_local_size(0);
+        const int sub_groups_in_work_group_num = work_group_size / sub_group_size; // num of subgroups for current node processing
 
-        __global const int * node = nodeList + nodeId * nNodeProp;
-        const int offset          = node[0];
-        const int nRows           = node[1];
-        const int featId          = node[2];
-        const int splitVal        = node[3];
-        const int nRowsLeft       = node[4]; // num of items in Left part
+        const int sub_group_local_id  = get_sub_group_local_id();
+        const int work_group_local_id = get_local_id(0);
 
-        if (featId != leafMark) // split node
+        const int sub_group_id   = get_group_id(0) * sub_groups_in_work_group_num + work_group_local_id / sub_group_size;
+        const int sub_groups_num = get_num_groups(0) * sub_groups_in_work_group_num; // num of subgroups for current node processing
+
+        const int totalBlocksNum = nNodes * maxBlocksNum;
+
+        for (int blockIndGlob = sub_group_id; blockIndGlob < totalBlocksNum; blockIndGlob += sub_groups_num)
         {
-            int sum = 0;
+            int nodeId   = blockIndGlob / maxBlocksNum;
+            int blockInd = blockIndGlob % maxBlocksNum;
+
+            __global const int * node = nodeList + nodeId * nNodeProp;
+            const int offset          = node[0];
+            const int nRows           = node[1];
+            const int featId          = node[2];
+            const int splitVal        = node[3];
+            const int nRowsLeft       = node[4]; // num of items in Left part of node
+
+            int nodeBlocks = nRows / minBlockSize ? min(nRows / minBlockSize, maxBlocksNum) : 1;
```

Comment:
```suggestion
            const int nodeBlocks = max(min(nRows / minBlockSize, maxBlocksNum), 1);
```

Reply:
in this case min will be called always, in my, only if condition is true